localstack/http/router.py

```diff
@@ -462,23 +373,223 @@ def route(
+        ])
+
+        router.add([
+            endpoints,
+            Submount("/foo", [endpoints])
+        ])
+
+    """
+
+    factory: RuleFactory
+    """The underlying real rule factory."""
+
+    @overload
+    def __init__(
+        self,
+        path: str,
+        endpoint: E,
+        host: Optional[str] = None,
+        methods: Optional[Iterable[str]] = None,
+        **kwargs,
+    ):
         """
-        DEPRECATED: use ``add`` instead.
+        Basically a ``Rule``.
+
+        :param path: the path pattern to match. This path rule, in contrast to the default behavior of Werkzeug, will be
+                        matched against the raw / original (potentially URL-encoded) path.
+        :param endpoint: the endpoint to invoke
+        :param host: an optional host matching pattern. if not pattern is given, the rule matches any host
+        :param methods: the allowed HTTP verbs for this rule
+        :param kwargs: any other argument that can be passed to ``werkzeug.routing.Rule``
+        """
+        ...
+
+    @overload
+    def __init__(self, fn: _RouteEndpoint):
+        """
+        Takes a route endpoint (typically a function decorated with ``@route``) and adds it as ``EndpointRule``.
+
+        :param fn: the RouteEndpoint function
+        """
+        ...
+
+    @overload
+    def __init__(self, rule_factory: RuleFactory):
+        """
+        Adds a ``Rule`` or the rules created by a ``RuleFactory`` to the given router. It passes the rules down to
+        the underlying Werkzeug ``Map``, but also returns the created Rules.
+
+        :param rule_factory: a `Rule` or ``RuleFactory`
         """
-        return self._add_route(fn)
+        ...
+
+    @overload
+    def __init__(self, obj: Any):
+        """
+        Scans the given object for members that can be used as a `RouteEndpoint` and adds them to the router.
 
-    def add_route_endpoints(self, obj: object) -> List[Rule]:
+        :param obj: the object to scan
         """
-        DEPRECATED: use ``add`` instead.
+        ...
+
+    @overload
+    def __init__(self, rules: list[Union[_RouteEndpoint, RuleFactory, Any]]):
+        """Add multiple rules at once"""
+        ...
+
+    def __init__(self, *args, **kwargs):
         """
-        return self._add_routes(obj)
+        Dispatcher for overloaded ``__init__`` methods.
+        """
+        if "path" in kwargs or type(args[0]) is str:
+            self.factory = _EndpointRule(*args, **kwargs)
+        elif "fn" in kwargs or callable(args[0]):
+            self.factory = _EndpointFunction(*args, **kwargs)
+        elif "rule_factory" in kwargs:
+            self.factory = kwargs["rule_factory"]
+        elif isinstance(args[0], RuleFactory):
+            self.factory = args[0]
+        elif isinstance(args[0], list):
+            self.factory = RuleGroup([RuleAdapter(rule) for rule in args[0]])
+        else:
+            self.factory = _EndpointsObject(*args, **kwargs)
```

Comment:
This is a nice encapsulation of the different possibilities of defining rules directly using Werkzeug classes! 💯